lib/application.js

```diff
@@ -398,85 +399,217 @@ app.enableAuth = function(options) {
     }
   };
 
-  this._verifyAuthModelRelations();
-
-  this.isAuthEnabled = true;
+  return this._verifyAuthModelRelations(app._warnOnBadAuthModelRelations)
+    .then(() => {
+      this.isAuthEnabled = true;
+    });
```

Comment:
I am quite concerned about this change. We use `app.isAuthEnabled` in other places (e.g. https://github.com/strongloop/loopback/blob/c3ab9320127cb7726e83b9668f641f7cb2a55692/server/middleware/rest.js#L51-L54), but possibly in any 3rd party LoopBack code to detect whether the app has authentication enabled. While our REST middleware makes this check few ticks later (at which time this promise chain has already finished), I think it's not safe to assume that everybody else is doing the same.

Having said that, I agree that making `enableAuth` async & promise returning is a good direction.

What is your reasoning for enabling `isAuthEnabled` only after auth-model relations have been verified? What would happen if we keep only `_verifyAuthModelRelations` async and set `app.isAuthEnabled` right on the first tick, as we are doing now?

BTW in the current proposal, the app will end up in an inconsistent state when `_verifyAuthModelRelations` -models and the remoting authorization hook are all set up, but the flag is saying otherwise.